cinder/api/metricutil.py

```
@@ -4,9 +4,110 @@
 @author: souvik
 '''
 from metrics.ThreadLocalMetrics import ThreadLocalMetrics, ThreadLocalMetricsFactory
+from metrics.Metrics import Unit
 from oslo_log import log as logging
+from time import time
+from oslo_context.context import RequestContext as context
 LOG = logging.getLogger(__name__)
 
+class ReportMetrics(object):
+    def __init__(self, metric_name, report_error = False):
+        self.__metric_name = metric_name
+        self.__report_error = report_error
+    def __call__(self, function):
+        def metrics_wrapper(*args, **kwargs):
+            start_time = time()
+            error = 0
+            try:
+                return function(*args, **kwargs)
+            except Exception as e:
+                LOG.exception('Exception in Report Metrics: %s', e)
+                error = 1
+            finally:
+                end_time = time()
+                try:
+                    metrics = ThreadLocalMetrics.get()
+                    metric_time = self.__metric_name + "_time"
+                    metrics.add_time(metric_time, int((end_time - start_time)*1000), Unit.MILLIS)
+                    if self.__report_error == True:
+                        metric_error = self.__metric_name + "_error"
+                        metrics.add_count(metric_error, error)
+                except Exception as e:
+                    LOG.exception("No threadlocal metrics object: %s", e)
```

If we know that the exception is due to absence of thread lcoal metrics object, then why not catch that rather than the generic exception?
Again, do we need to reraise the exception?